src/overlords/situational/stronghold.ts

```diff
@@ -0,0 +1,316 @@
+import {log} from '../../console/log';
+import {CreepSetup} from '../../creepSetups/CreepSetup';
+import {CombatSetups, Roles} from '../../creepSetups/setups';
+import {DirectiveInvasionDefense} from '../../directives/defense/invasionDefense';
+import {DirectiveStronghold} from '../../directives/situational/stronghold';
+import {CombatIntel} from '../../intel/CombatIntel';
+import {OverlordPriority} from '../../priorities/priorities_overlords';
+import {profile} from '../../profiler/decorator';
+import {boostResources} from '../../resources/map_resources';
+import {derefCoords, getCacheExpiration, getPosFromString} from '../../utilities/utils';
+import {Visualizer} from '../../visuals/Visualizer';
+import {CombatZerg} from '../../zerg/CombatZerg';
+import {CombatOverlord} from '../CombatOverlord';
+
+
+
+/**
+ * Prioritized list of what order enemy structures should be attacked in
+ */
+export const StrongholdAttackPriorities: StructureConstant[] = [
+	STRUCTURE_INVADER_CORE,
+	STRUCTURE_TOWER,
+	STRUCTURE_RAMPART,
+	STRUCTURE_WALL,
+];
+
+/**
+ * Spawns ranged attacker against stronghold
+ */
+@profile
+export class StrongholdOverlord extends CombatOverlord {
+
+	strongholdKillers: CombatZerg[];
+	room: Room | undefined;
+	directive: DirectiveStronghold;
+	_target: RoomObject | null;
+	_attackPos?: RoomPosition;
+
+	static settings = {
+		retreatHitsPercent : 0.85,
+		reengageHitsPercent: 0.95,
+	};
+
+	constructor(directive: DirectiveStronghold,
+				priority = OverlordPriority.defense.rangedDefense) {
+		super(directive, 'stronghold', priority, 1);
+		this.strongholdKillers = this.combatZerg(Roles.strongholdKiller, {
+			notifyWhenAttacked: false,
+			boostWishlist: [boostResources.tough[3], boostResources.ranged_attack[3],
+					boostResources.heal[3], boostResources.move[3]]
+		});
+	}
+
+	/**
+	 * Returns [position, target] by searching for safe locations
+	 * @param target
+	 * @param range
+	 * @param myCreep
+	 */
+	private findAttackingPositionAndTarget(target: Creep | Structure, range: number,
+										   myCreep: CombatZerg): {attackPos: RoomPosition, target: Creep | Structure} | undefined {
+		log.info(`Finding attacking position in ${target.room} for ${this.print}`);
+		if (!target.room || range == 0) {return;}
+
+		// bug with containers
+		const shootPositions = target.pos.getPositionsAtRange(range, false, false);
+
+		// Index is range to avoid from
+		const avoidLocs: RoomPosition[][] = Array.from({length: 5}, () => []);
+
+		avoidLocs[1] = avoidLocs[1].concat(target.room.ramparts.map(rampart => rampart.pos));
+		avoidLocs[4] = avoidLocs[4].concat(target.room.sources.map(s => s.pos));
+		avoidLocs[4] = avoidLocs[4].concat(target.room.keeperLairs.map(s => s.pos));
+		if (target.room.mineral) {avoidLocs[4] = avoidLocs[4].concat(target.room.mineral.pos);}
+
+		// Array where first index is how many hostile locations there are
+		const safeSpots: RoomPosition[][] = this.findSafeLocation(shootPositions, avoidLocs);
```

Comment:
make this an object with keys `shootPositions` and `avoidPositions`... array access like `thing[0]` is ugly